### DISCUSSION TITLE
So there are some things I would love to see.

### DISCUSSION THREAD

**Discussion by agluck91** [blinksh/blink#1824] [Ideas]
On Mac, I love using warp shell. There are some things on this that I feel would be great additions to blink. Blink is already in a class of its own, completely unrivaled, but it would further set it apart it if had these features. 

1. Warp just introduced warp drive. This allows you to create injectable templatized commands. So for instasntance, I create pulumi secrets all the time. I made a warp drive entry for it, I click on it and it auto populates the command into the input buffer. This could easily be something you type and not click, in fact a have a feature request to make warp drive more assessable to keyboard users. I rarely touch my mouse.
So this would have to be baked into the shell itself, it is not something that will work as a program on the remote host. Picture bellow of an example. It would be nice to have like a power sidebar. I know you are going for the clean look, so maybe it can be a swipe up feature, as of right now swipe up with one finger does nothing. 
![image](https://github.com/blinksh/blink/assets/113385018/2308525b-e81d-44cc-a901-d6c38e3c1af9)

2. Warp is intelligently auto completing commands. I am not sure how hard this is to do, but man would it be nice to tab complete commands like you can with warp. Most of the time it looks like it is pulling from my history. all you have to do is press right arrow to accept the entry. 80% of the time it is very helpful. 
![image](https://github.com/blinksh/blink/assets/113385018/3615cf95-c92d-4776-907d-3db4d4b50efe)

3. this is not in warp, but a neat idea, and this would be a TON of work, but it would be cool if a text editor could be added to warps commands. I am a nvim fan, but any text editor would work, and then have it callable with the share screen, and allow me to open files in blink. Right now I am using a different editor, but it would be cool to have something vim like for local editing

**Comment by carloscabanero** [comment]
I think we have already been working on this:
1. Have you checked Snips? It will be out tomorrow https://docs.blink.sh/basics/snips
2. I think this is more a task for your shell of choice (zsh or fish already do this). But, we are planning to add History on Snips, so you can easily repeat the previous sequence, etc...
3. We already embedded Runestone for Snips inside Blink, next step would be to add more editor functionality to it, but we need to be clear on the scenario. For example, we have a *scratch* mode in Snips so you can edit commands before pasting on shell, etc... The goal is that snips can work in multiple languages too, etc...

**Reply by agluck91** [reply]
1. Woah!! I missed snips completely. I did not see the announcement! That is exactly what I am looking for, and I think you did a better implementation of it quite frankly. 

I love blink so much that I typically use it as my main terminal even with my Mac with universal control. I wish it would run properly on the m1 Mac, but I think that is asking too much. I tried running it on the Mac, but it just did weird things. 



2. I know the shell can do that and maybe I need to look at some plugins for that, warp has it built into the input prompt. And they use their ai library to enhance it. I am a zsh fan, but mine does not seem to do it out of the box on blink. So I must be missing something there. I will look at the plugins again. 

3. I think that will be very powerful. I think the scratch pad would have been sufficient for me. However, A scenario is, the other night I needed to open an sql file and manipulate its contents. I had the file on my iPad, because I was working solely from my iPad that night.I had to leave blink to do everything then can back and pasted the contents in over ssh. If I had a local vim that the share sheet knew about it would have been easier. Or maybe the built in vscode could do that. I have just been having problems with the embedded vscode a lot lately; After the last couple of updates ago. I have to switch it to pre release to get any of my settings to work right including connecting to tunnels.

**Reply by carloscabanero** [reply]
1. Thanks! It may look simple, but Yury and I put a lot of thinking on the mechanics of how it should work and how we could continue to extend it. It feels very natural and fast. I use it now all the time for a lot of things, from connecting to my remote and continuing a specific tmux session, to keeping our "deployment" commands in one place. We do have a lot of other ideas for how to continue extending it.
2. Our idea here is that "autocompletion" should happen from the shell itself. People have different priorities there and we have no intention to compete with Zsh, etc... But Snips fit into this as we extend with more Indexes. For example, there could be an Index for Git, or Rust, or Pulumi even, and it could serve as a quick guide that then you can further customize for your usage. And obviously there is AI...
3. For this use case, I would have created a Snip (although we do not allow different extensions other than shell, we will add this ability soon), and then you can edit the Snip all you want and paste or send to a terminal whenever you need it. Delete it once done. We think a Snip is perfect for this short pieces of code, which are not part of a codebase per-se, and that you may need to run more often. We are also thinking about ways to share them over GitHub. Check out #1755.

About code, there are a few issues we are aware of but we can only go that far with our fixes. Take into account Blink Code is just a facade over a remote VS Code session. We are trying to figure out what the best path forward for this is.

**Reply by agluck91** [reply]
I can tell, When you have a function that is that seamless, you always have a substantial amount of work in designing it. 

Yury and you did a GREAT job on this. I am excited. I see a lot of potential here. I will get ZSH to do what I want it to. 
I am excited to get in on this. 

So question, blink 14 run on Mac M1 with quite a bit of trouble, but it does run. How hard do you think it would be to bring blink to the MacBook. I know that you designed this specifically for iPad, but with the bridging they have introduced for iPad programs on Mac it may be feasible! I would drop warp in a heartbeat lol. It would only be possible on M1 Macs, so I know that puts a lot of people out, but if it is a simple of a flip of a switch it may be nice!!

**Reply by carloscabanero** [reply]
At the moment it is not just flipping a switch. We need to evaluate this in two stages: first to just make it work in a similar manner to the current app, then a second step would be to offer something that actually makes sense on the Mac (ie, the initial shell does not make sense there). Any feedback on what Blink for the Mac should be is really welcome!

There is also a thread here, but not a lot of conversation https://github.com/blinksh/blink/discussions/1627

**Reply by agluck91** [reply]
Okay. I will reply in that thread and close this one now!
